### ISSUE TITLE
Weird type-safety behaviour when using $ helper

### ISSUE THREAD

**Issue by Ivanrenes** [kossnocorp/typesaurus#132]
Hey I'm having an issue with type safety when using $ helper.

Seems like when using $ helper, we lost full type safety since TS does not complain about "inexistentProp".
![image](https://github.com/kossnocorp/typesaurus/assets/34603733/9fa6aba0-ad7e-4d6a-8c91-b3498de021dd)

However if we remove the $ helper, it works as expected, triggering "inexistentProp" as not part of type User
![image](https://github.com/kossnocorp/typesaurus/assets/34603733/21b285d2-469c-493e-bdb3-20038899e85f)

Versions tested:
- 10.3.0
- 10.7.0

If need more context lmk!

**Comment by kossnocorp** [comment]
Hey! Thank you for opening the issue. Here's what's going on:

The problem is that TypeScript has structured typing, and there's no way to fix it without making the developer experience slightly worse.

Here's [a playground link with the gist of the problem](https://www.typescriptlang.org/play/?#code/JYOwLgpgTgZghgYwgAgKoGdrIN4ChnIzBTpgBycAthAFzKlSgDmA3PsgDZykXV0PM2BOB2Dd+YRiFa4AvrlwwAriARhgAexDIlABwAmcSLwgAKEFVrIACsAQBrADwZoAGmQAiIiXKWPyAB9PLh4-AD4AShx5XD1DY0tTPAJvUL5PAGVuAAs4D1d2EN90jwBpDXR0fPYAehrkQF4NwBkd5ABRAA9dCDUIfWRoKA0odwB3FH0tAHIwZBG4cGQwDWQNADdoEcZIfvbJOEJgCA59dHYRMXQ6DxANLurZCLZcBC1SZF04AE8ODTg+gF4cOxUsUrB4suhctUCEUTFdypVochzuJPDc7gVZE86o1ALR7yDIywGQ3cYGyKDAny6s2A+ggIGYyLUwFWRl6sQMbJMpg+31++keuCAA).

And here's one of many TypeScripts related to it: https://github.com/microsoft/TypeScript/issues/12936

TL;DR: TypeScript can't do it.

However, I found a solution that works, but as I said, it makes the experience slightly worse. I can't ship it with the v10 as it will be a breaking change, and also, I'm hesitant to introduce it as it will break so much code even though the user-code fix would be pretty trivial 😞

So here's what I did in [Superstate](https://github.com/kossnocorp/superstate), my other type-safe project:

```ts
form.send.submit("error", "-> errored", ($, context) =>
  $({ ...context, error: "Email is missing" })
);
```

Instead of returning an object, it forces to wrap it into an argument. It triggers an extra properties check that is missing from the function return but present on argument passing.

So, in Typesaurus, it would look like this:

```ts
await db.users.add(($) => $({
  name: "Sasha",
  createdAt: $.serverDate(),
  inexistentProp: ''
  //^^^^^^^^^^^^
  // Type error
}));
```

I think it's worth it, but as I said, it will require a breaking change, and I wanted to keep v10 stable for a long time. I could use a `schema` function option to switch types used for the write helpers, but I'm unsure how much work it would be and if it's worth the sacrifice to preserve v10 stability. 

What do you think? I would appreciate any feedback.
